File: queens-attack-2/solution.clj
;--------------------------------------------------
; Calculate the distance between a and b
;
; It's the distance on the chess board, so 4 and 3
; are next to each other and therefore have a distance
; of zero.
;--------------------------------------------------
(defn delta [a b]
  (dec (Math/abs (- a b))))

;--------------------------------------------------
; Given the position of the queen and the position
; of the obstacles, find the closest obstacle for
; each of the 8 directions
;--------------------------------------------------
(defn get-nearest [queen obstacles]
  (let [queen-row (nth queen 0) queen-col (nth queen 1) near {:N -1 :S -1 :E -1 :W -1 :NE -1 :NW -1 :SE -1 :SW -1}]
    (reduce (fn [near ob]
              (let [ob-col (nth ob 1)
                    ob-row (nth ob 0)
                    delta-col (delta queen-col ob-col)
                    delta-row (delta queen-row ob-row)
                    is-diag (= delta-col delta-row)]
                (if is-diag
                  (cond
                    (and (< ob-row queen-row) (< ob-col queen-col)
                         (or (= -1 (near :SW)) (< delta-col (near :SW)))) (assoc near :SW delta-col)
                    (and (< ob-row queen-row) (> ob-col queen-col)
                         (or (= -1 (near :SE)) (< delta-col (near :SE)))) (assoc near :SE delta-col)
                    (and (> ob-row queen-row) (< ob-col queen-col)
                         (or (= -1 (near :NW)) (< delta-col (near :NW)))) (assoc near :NW delta-col)
                    (and (> ob-row queen-row) (> ob-col queen-col)
                         (or (= -1 (near :NE)) (< delta-col (near :NE)))) (assoc near :NE delta-col)
                    :else near)
                  (cond
                    (and (= ob-col queen-col) (< ob-row queen-row)
                         (or (= -1 (near :S)) (< delta-row (near :S))))  (assoc near :S delta-row)
                    (and (= ob-col queen-col) (> ob-row queen-row)
                         (or (= -1 (near :N))  (< delta-row (near :N))))  (assoc near :N delta-row)
                    (and (= ob-row queen-row) (< ob-col queen-col)
                         (or (= -1 (near :W)) (< delta-col (near :W)))) (assoc near :W delta-col)
                    (and (= ob-row queen-row) (> ob-col queen-col)
                         (or (= -1 (near :E)) (< delta-col (near :E)))) (assoc near :E delta-col)
                    :else near))))
            near
            obstacles)))

;--------------------------------------------------
; Given the closest obstacle in each direction,
; calculate the total squares the queen can attack
;--------------------------------------------------
(defn calc-squares [sideLength queen near]
  (let [queen-row (nth queen 0) queen-col (nth queen 1) total 0]
    (reduce (fn [total near]
              (if (= (near 1) -1)
                (case (near 0)
                  :N (+ total (delta (inc sideLength) queen-row))
                  :S (+ total (inc (delta 1 queen-row)))
                  :E (+ total (delta (inc sideLength) queen-col))
                  :W (+ total (inc (delta 1 queen-col)))
                  :NE (+ total (Math/min (delta (inc sideLength) queen-col) (delta (inc sideLength) queen-row)))
                  :NW (+ total (Math/min (inc (delta 1 queen-col)) (delta (inc sideLength) queen-row)))
                  :SE (+ total (Math/min (delta (inc sideLength) queen-col) (inc (delta 1 queen-row))))
                  :SW (+ total (Math/min (inc (delta 1 queen-col)) (inc (delta 1 queen-row)))))
                (+ total (near 1))))
            total
            near)))

;--------------------------------------------------
; Real work starts here
;--------------------------------------------------
(def sizes (map #(Integer/parseInt %) (clojure.string/split (read-line) #" ")))
(def sideLength (nth sizes 0))
(def numObstacles (nth sizes 1))
(def queen (map #(Integer/parseInt %) (clojure.string/split (read-line) #" ")))
(def obstacles (for [_ (range numObstacles)]
                 (map #(Integer/parseInt %) (clojure.string/split (read-line) #" "))))

(println (calc-squares sideLength queen (get-nearest queen obstacles)))

File: queens-attack-2/solution.py
import math
import os
import random
import re
import sys

# Complete the queensAttack function below.


def delta(a, b):
    return abs(a - b) - 1


def printGrid(n, r_q, c_q, obstacles):
    grid = [['-' for x in range(n + 1)] for x in range(n + 1)]

    for i in range(n + 1):
        if i == 0:
            continue
        for j in range(n + 1):
            if j == 0:
                continue
            if i == r_q or j == c_q:
                grid[i][j] = '*'
            if delta(r_q, i) == delta(c_q, j):
                grid[i][j] = '*'

    grid[r_q][c_q] = 'Q'
    for ob in obstacles:
        grid[ob[0]][ob[1]] = 'X'

    for i in range(n + 1):
        if i == 0:
            continue
        for j in range(n + 1):
            if j == 0:
                continue
            print(grid[i][j], end='')
        print()


def queensAttack(n, k, r_q, c_q, obstacles):
    # printGrid(n, r_q, c_q, obstacles)

    N = 0
    S = 1
    E = 2
    W = 3
    NW = 4
    NE = 5
    SW = 6
    SE = 7

    names = ['N', 'S', 'E', 'W', 'NW', 'NE', 'SW', 'SE']
    near = [-1, -1, -1, -1, -1, -1, -1, -1]

    for ob in obstacles:
        deltac = delta(ob[1], c_q)
        deltar = delta(ob[0], r_q)

        if ob[0] < r_q:
            if ob[1] == c_q:
                if near[S] == -1 or delta(r_q, ob[0]) < near[S]:
                    near[S] = deltar

            elif delta(ob[0], r_q) == delta(ob[1], c_q):
                value = deltac

                if ob[1] > c_q and (near[SE] == -1 or value < near[SE]):
                    near[SE] = value

                if ob[1] < c_q and (near[SW] == -1 or value < near[SW]):
                    near[SW] = value

        elif ob[0] > r_q:
            if ob[1] == c_q:
                if near[N] == -1 or delta(r_q, ob[0]) < near[N]:
                    near[N] = deltar

            elif delta(ob[0], r_q) == delta(ob[1], c_q):
                value = deltac
                if ob[1] > c_q and (near[NE] == -1 or delta(ob[1], c_q) < near[NE]):
                    near[NE] = value
                if ob[1] < c_q and (near[NW] == -1 or delta(ob[1], c_q) < near[NW]):
                    near[NW] = value

        if ob[1] < c_q:
            if ob[0] == r_q:
                if near[W] == -1 or delta(c_q, ob[1]) < near[W]:
                    near[W] = deltac

        elif ob[1] > c_q:
            if ob[0] == r_q:
                if near[E] == -1 or delta(c_q, ob[1]) < near[E]:
                    near[E] = deltac

    total = 0
    for i in range(len(near)):
        if near[i] != -1:
            # print('near{} {}'.format(names[i], near[i]))
            total += near[i]

    if near[N] == -1:
        # print('N {}'.format(delta(n + 1, r_q)))
        total += delta(n + 1, r_q)

    if near[S] == -1:
        # print('S {}'.format(delta(1, r_q) + 1))
        total += delta(1, r_q) + 1

    if near[E] == -1:
        # print('E {}'.format(delta(n + 1, c_q)))
        total += delta(n + 1, c_q)

    if near[W] == -1:
        # print('W {}'.format(delta(1, c_q) + 1))
        total += delta(1, c_q) + 1

    if near[NE] == -1:
        # print('NE {}'.format(min(delta(n + 1, c_q), delta(n + 1, r_q))))
        total += min(delta(n + 1, c_q), delta(n + 1, r_q))

    if near[NW] == -1:
        # print('NW {}'.format(min(delta(1, c_q) + 1, delta(n + 1, r_q))))
        total += min(delta(1, c_q) + 1, delta(n + 1, r_q))

    if near[SE] == -1:
        # print('SE {}'.format(min(delta(n + 1, c_q), delta(1, r_q))))
        total += min(delta(n + 1, c_q), delta(1, r_q) + 1)

    if near[SW] == -1:
        # print('SW {}'.format(min(delta(1, c_q) + 1, delta(1, r_q) + 1)))
        total += min(delta(1, c_q) + 1, delta(1, r_q) + 1)

    return total


if __name__ == '__main__':
    nk = input().split()

    n = int(nk[0])

    k = int(nk[1])

    r_qC_q = input().split()

    r_q = int(r_qC_q[0])

    c_q = int(r_qC_q[1])

    obstacles = []

    for _ in range(k):
        obstacles.append(list(map(int, input().rstrip().split())))

    result = queensAttack(n, k, r_q, c_q, obstacles)

    print(result)
